projectq/cengines/_optimize_test.py

```diff
@@ -147,3 +226,488 @@ def test_local_optimizer_identity_gates():
     # Expect allocate, one Rx gate, and flush gate
     assert len(backend.received_commands) == 3
     assert backend.received_commands[1].gate == Rx(0.5)
+
+def test_local_optimizer_commutable_gates():
+    # Test that inverse gates separated by two commutable gates 
+    # cancel successfully and that mergeable gates separated by
+    # two commutable gates cancel successfully.
+    local_optimizer = _optimize.LocalOptimizer(m=5)
+    backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=backend, engine_list=[local_optimizer])
+    qb0 = eng.allocate_qubit()
+    qb1 = eng.allocate_qubit()
+    qb2 = eng.allocate_qubit()
+    qb3 = eng.allocate_qubit()
+    Rx(0.4) | qb0
+    Rx(-math.pi) | qb1
+    Rxx(0.3) | (qb1, qb0)
+    Rxx(0.5) | (qb0, qb1)
+    Rx(math.pi) | qb1
+    Rx(0.5) | qb0
+    Rxx(0.2) | (qb2, qb3)
+    Rx(0.3) | qb2
+    Rx(0.2) | qb3
+    Rxx(0.2) | (qb3, qb2)
+    assert len(backend.received_commands) == 0
+    eng.flush()
+    received_commands = []
+    # Remove Allocate and Deallocate gates
+    for cmd in backend.received_commands:
+        if not (isinstance(cmd.gate, FastForwardingGate) or
+                isinstance(cmd.gate, ClassicalInstructionGate)):
+            received_commands.append(cmd)
+    assert len(received_commands) == 5
+    assert received_commands[0].gate == Rx(0.9)
+    # If this test doesn't succeed check that the Rxx 
+    # interchangeable qubits attribute is working.
+    assert received_commands[1].gate == Rxx(0.8)
+    assert received_commands[2].gate == Rxx(0.4)
+    assert received_commands[3].gate == Rx(0.3)
+    assert received_commands[4].gate == Rx(0.2)
+    assert received_commands[0].qubits[0][0].id == qb0[0].id
+    assert received_commands[1].qubits[0][0].id == qb0[0].id
+    assert received_commands[1].qubits[1][0].id == qb1[0].id
+    assert received_commands[2].qubits[0][0].id == qb2[0].id
+    assert received_commands[2].qubits[1][0].id == qb3[0].id
+    assert received_commands[3].qubits[0][0].id == qb2[0].id
+    assert received_commands[4].qubits[0][0].id == qb3[0].id
+
+def test_local_optimizer_commutable_circuit_Rz_example_1():
+    # Rzs should merge
+    local_optimizer = _optimize.LocalOptimizer(m=10)
+    backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=backend, engine_list=[local_optimizer])
+    qb0 = eng.allocate_qubit()
+    qb1 = eng.allocate_qubit()
+    Rz(0.1) | qb0
+    H | qb0
+    CNOT | (qb1, qb0)
+    H | qb0
+    Rz(0.2) | qb0
+    eng.flush()
+    received_commands = []
+    # Remove Allocate and Deallocate gates
+    for cmd in backend.received_commands:
+        if not (isinstance(cmd.gate, FastForwardingGate) or
+                isinstance(cmd.gate, ClassicalInstructionGate)):
+            received_commands.append(cmd)
+    assert received_commands[0].gate == Rz(0.3)
+    assert len(received_commands) == 4
+
+def test_local_optimizer_commutable_circuit_Rz_example_2():
```

Comment:
Same comment as above about making this test parameterised.

Reply:
Done. As will be seen in next commit.